### DISCUSSION TITLE
Permission seeder - PermissionDoesNotExist

### DISCUSSION THREAD

**Discussion by basic-ph** [spatie/laravel-permission#2708] [Q&A] [answered]
Hello hello! 👋 

This is the first time using `laravel-permission` so maybe I'm missing something important.
I'm experiencing a strange issue while seeding my database with permissions and roles.

This is the code in my `DatabaseSeeder.php` which is very close to the template default:
```php
<?php

declare(strict_types=1);

namespace Database\Seeders;

use App\Models\Material;
use App\Models\User;
use Illuminate\Database\Console\Seeds\WithoutModelEvents;
use Illuminate\Database\Seeder;

final class DatabaseSeeder extends Seeder
{
    use WithoutModelEvents;

    /**
     * Seed the application's database.
     */
    public function run(): void
    {

        $this->call([
            RolesAndPermissionsSeeder::class,
        ]);

        // User::factory(10)->create();

        User::factory()->create([
            'name' => 'Test User',
            'email' => 'test@example.com',
        ]);

        // Material
        Material::factory(3)->create();
    }
}
```
 
 The `RolesAndPermissionsSeeder.php` file code is the one provided by documentation [Roles/Permissions Seeder](https://spatie.be/docs/laravel-permission/v6/advanced-usage/seeding#content-rolespermissions-seeder).
 
 In this situation when I run the command `./vendor/bin/sail artisan migrate:fresh --seed` I get the following error:
 ```sh
 [...]
    INFO  Seeding database.  

  Database\Seeders\RolesAndPermissionsSeeder ............................................................................................... RUNNING  

   Spatie\Permission\Exceptions\PermissionDoesNotExist 

  There is no permission named `edit articles` for guard `web`.

  at vendor/spatie/laravel-permission/src/Exceptions/PermissionDoesNotExist.php:11
      7▕ class PermissionDoesNotExist extends InvalidArgumentException
      8▕ {
      9▕     public static function create(string $permissionName, ?string $guardName)
     10▕     {
  ➜  11▕         return new static("There is no permission named `{$permissionName}` for guard `{$guardName}`.");
     12▕     }
     13▕ 
     14▕     /**
     15▕      * @param  int|string  $permissionId

      +6 vendor frames 

  7   database/seeders/RolesAndPermissionsSeeder.php:26
      Spatie\Permission\Models\Role::givePermissionTo()
      +8 vendor frames 

  16  database/seeders/DatabaseSeeder.php:22
      Illuminate\Database\Seeder::call()
 ```
 
 While trying to understand the issue I made some tests. I noticed the following behavior: If I store the first `Permission` model instance inside a variable and later use that variable to pass it as argument to `givePermissionTo` method the seed works fine.
 I checked also in my database tables and everything ***seems*** right to me.
This is the "working" code:
```php
<?php

namespace Database\Seeders;

use Illuminate\Database\Seeder;
use Spatie\Permission\Models\Role;
use Spatie\Permission\Models\Permission;

class RolesAndPermissionsSeeder extends Seeder
{
    public function run(): void
    {
        // Reset cached roles and permissions
        app()[\Spatie\Permission\PermissionRegistrar::class]->forgetCachedPermissions();

        // create permissions
        $permission1 = Permission::create(['name' => 'edit articles']); // PERMISSION STORED IN VARIABLE
        Permission::create(['name' => 'delete articles']);
        Permission::create(['name' => 'publish articles']);
        Permission::create(['name' => 'unpublish articles']);

        // create roles and assign created permissions

        // this can be done as separate statements
        $role = Role::create(['name' => 'writer']);
        $role->givePermissionTo($permission1); // VARIABLE USED TO LINK PERMISSION TO ROLE

        // or may be done by chaining
        $role = Role::create(['name' => 'moderator'])
            ->givePermissionTo(['publish articles', 'unpublish articles']);

        $role = Role::create(['name' => 'super-admin']);
        $role->givePermissionTo(Permission::all());
    }
}
``` 

Any clarification or explanation of this behavior would be enormously helpful?
Thank you in advance 🚸 🆘

**Comment by drbyte** [comment]
The examples don't specify a guard (because they're intended to be more generic), so it defaults to the first guard defined in your `/config/auth.php` list of guards. If you've changed that from `'web'` to something else, that could explain why it might not be finding the record.

But, as you've done, assigning the created record to a variable, and using that later is a perfectly acceptable approach.

**Reply by basic-ph** [reply]
I have not changed the default guard, it is currently “web”.
I am wondering why it is not possible to assign a permission to a role using the `name` as stated in the documentation for the first role (`writer`) but it seems perfectly fine for the 2nd (`moderator`) role.

**Reply by drbyte** [reply]
I just created a new project to test it, and ran the example seeder without any problems:

```sh
laravel new permissionseed
```
(Chose No UI, No git, SQLite, Migrate:yes)
```sh
composer require spatie/laravel-permission
php artisan vendor:publish
```
(chose Spatie Permissions package from the Publish list)
```sh
php artisan migrate
php artisan make:seed RolesAndPermissionsSeeder
```
Edited the `RolesAndPermissionsSeeder.php` file as shown in the docs page you linked to.
Also edited `DatabaseSeeder.php` to add `$this->call([RolesAndPermissionsSeeder::class]);` as you had done above.
```sh
php artisan db:seed
```
Seeded without issues.

Cleared and ran again to test:
```sh
php artisan migrate:fresh --seed
```
Seeded without issues.

Repeated with the seeder example in https://spatie.be/docs/laravel-permission/v6/basic-usage/new-app as well, for thoroughness. Seeded, but with one problem: the User seeding gave an error because both seeders use `test@example.com` for the user email. Easy change. Then migrate:fresh --seed again, without problems.  (EDIT: also had to add the `HasRoles` trait to the `User` model, as per normal install of this package.)

What's different in the process you've been using?

**Reply by drbyte** [reply, marked as answer]
OH ... Maybe you're using `use WithoutModelEvents;`?

Try without that.

Or add a cache-flush after creating those permissions:

```diff
    use WithoutModelEvents;

    public function run(): void
    {
        // Reset cached roles and permissions
        app()[PermissionRegistrar::class]->forgetCachedPermissions();

        // create permissions
        Permission::create(['name' => 'edit articles']);
        Permission::create(['name' => 'delete articles']);
        Permission::create(['name' => 'publish articles']);
        Permission::create(['name' => 'unpublish articles']);

        // update cache
+        app()[PermissionRegistrar::class]->forgetCachedPermissions();

```

**Reply by basic-ph** [reply]
Thank you very much @drbyte for your help!
The problem was that I have `WithoutModelEvents` trait in my `DatabaseSeeder` class. My experience with Laravel is still quite limited, I thought that the trail would affect the behaviour of DatabaseSeeder only but as it says here: 

> When used, the WithoutModelEvents trait ensures no model events are dispatched, even if additional seed classes are executed via the call method

[Muting Model Events](https://laravel.com/docs/11.x/seeding#muting-model-events)

In an empty project in Laravel the Database seeder class has the default trait WithoutModelEvents. Do you think it might be useful to include a commented line in the example of `RolesAndPermissionsSeeder.php` ?

Something like this:
```php
// Reset cached roles and permissions
app()[\Spatie\Permission\PermissionRegistrar::class]->forgetCachedPermissions();

// create permissions
Permission::create(['name' => 'edit articles']);
Permission::create(['name' => 'delete articles']);
Permission::create(['name' => 'publish articles']);
Permission::create(['name' => 'unpublish articles']);

// Update cache if the WithoutModelEvents trait is used here or in DatabaseSeeder.php
app()[\Spatie\Permission\PermissionRegistrar::class]->forgetCachedPermissions();

// create roles and assign created permissions
[...]
```

**Reply by drbyte** [reply]
Agreed. Docs updated in the repo. The docs site will update on next rebuild/push.
